### PR TITLE
[CMAKE] Flexify and Bisonify automagically at build time. Needs flex and bison to be available (which RosBE provides).

### PR DESCRIPTION
**Pull request by AmineKhaldi**
CORE-15991

### CODE REVIEW

File: dll/directx/wine/d3dcompiler_43/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,9 @@ add_definitions(
 include_directories(${REACTOS_SOURCE_DIR}/sdk/include/reactos/wine)
 spec2def(d3dcompiler_43.dll d3dcompiler_43.spec ADD_IMPORTLIB)
 
+add_bison_files(asmshader.y hlsl.y ppy.y)
+add_flex_files(asmshader.l hlsl.l ppl.l)
```
[the remark this answers is not in the document]
Comment (AmineKhaldi) [reply dll/directx/wine/d3dcompiler_43/CMakeLists.txt]: They are, by definition, bound to the target they've added as source files to, so yes, it works as you would like it to.

**Comment by ThFabba** [comment]
Wooo!

File: dll/directx/wine/d3dcompiler_43/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimtabor) [reply dll/directx/wine/d3dcompiler_43/CMakeLists.txt]: Linux distribution (Fedora/Slackware/...) provides these already.

**Comment by SergeGautherie** [comment]
msvc-x64: no bison, no flex?

**Comment by ThFabba** [comment]
RosBE-Windows includes bison and flex, so this is not an issue on Windows.

**Comment by SergeGautherie** [comment]
> this is not an issue on Windows

What about ' AppVeyor build failed', then?
Is `vssolution` expected to behave any better?

**Comment by HBelusca** [comment]
> What about ' AppVeyor build failed', then?

Iirc these bots use a "stripped down" version of rosbe. (You can actually check what they do: first line says
> Extracting archive: C:\RosCMakeNinja.zip

suggesting it's just using cmake+ninja only without any other 3rd-party tool, i.e. it's not the complete RosBE).

**Comment by SergeGautherie** [comment]
> cmake+ninja only

Yes, I know that.

--

I was asked to test whether this PR fixes CORE-15991 (VSS build, ...),
but it looks like it does the opposite, as in breaks MSVC builds (too),
so I am rather confused...

**Comment by tkreuzer** [comment]
I'm opposed to this change, as it adds more dependencies to the build-environment and just because some stuff is already in RosBE, doesn't mean it's good to use it. For MSVC build the dependencies are VS+CMake+Ninja. Now you have to install the full RosBE and put it into the path environment variable to be able to use it from VS command prompt. This can create conflicts with other installations of CMake, Ninja, or any of the stuff in there.
And I don't see what it fixes.

**Comment by ThFabba** [comment]
What it fixes is we're one step closer to actually building Wine modules from their source code, rather than having to build Wine and then copy files that are generated as part of the build process. This not only makes syncs easier, it also makes them more likely to be correct, as it's very easy to forget updating these files (the .y/.l rarely change).

As far as I'm concerned, you were always supposed to do MSVC builds with `RosBE\bin` in your PATH, and using a custom CMake/Ninja was a do-it-at-your-own-risk option. It's not a coincidence that these tools are already included in RosBE. They're standard build tools used by many projects, and I'd imagine sooner or later we're likely to even want them for ROS-specific stuff.

I strongly feel this is a step in the right direction (ideally we'd get rid of *.rgs too). I'd argue that checking in generated files was just a hack in the first place.

**Comment by ColinFinck** [comment]
I'm all for this provided that we agree on specific Bison and Flex versions for RosBE 2.2 (and add a scripted build process to buildtoolchain.sh).
Currently, we rely on whatever version the OS ships on RosBE-Unix while RosBE-Windows just includes an unspecified Win32 version.

**Comment by SergeGautherie** [comment]
Doesn't `base/applications/winhlp32/CMakeLists.txt` need to be updated too? See 09c4d0a74b2362e2c9d29ec87c54ecc3bddd5b79.

**Comment by ColinFinck** [comment]
I have packaged Bison 3.5 and Flex 2.6.4+ for RosBE and modified the build scripts to include them.
Updated binaries and sources are here: https://svn.reactos.org/RosBE-Temp/
Remaining changes are in the RosBE Git repo.

**Comment by SergeGautherie** [comment]
Is there any hope to complete this (build regression) PR for v0.4.13?
Or does it depend on future ROSBE 2.2?

**Comment by ThFabba** [comment]
What makes you think this is a regression? As far as I know it's a new feature.
And yes, we should wait for the new RosBE.
And no, it's much too big a change for no user-visible (only developer-facing) benefit, so it wouldn't go in a release branch.

**Comment by SergeGautherie** [comment]
> regression ... new feature

Using flex and bison is new, yet CORE-15991 is a regression.

> wait for the new RosBE

Noted. I simply asked, because the Jira issue is/was still targeted at 0.4.13.

**Comment by Extravert-ir** [comment]
We've got some warnings:
```
 /home/runner/work/reactos/reactos/src/sdk/tools/widl/parser.y:323.1-14: warning: deprecated directive: ‘%error-verbose’, use ‘%define parse.error verbose’ [-Wdeprecated]
  323 | %error-verbose
      | ^~~~~~~~~~~~~~
      | %define parse.error verbose
/home/runner/work/reactos/reactos/src/sdk/tools/widl/parser.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
```
```
 /home/runner/work/reactos/reactos/src/dll/win32/jscript/parser.y: warning: 1 shift/reduce conflict [-Wconflicts-sr]
/home/runner/work/reactos/reactos/src/dll/win32/jscript/parser.y: warning: 18 reduce/reduce conflicts [-Wconflicts-rr]
```
```
 /home/runner/work/reactos/reactos/src/dll/win32/msi/cond.y:119.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
  119 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
/home/runner/work/reactos/reactos/src/dll/win32/msi/cond.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
[11584/12659] Generating sql.tab.c, sql.tab.h
/home/runner/work/reactos/reactos/src/dll/win32/msi/sql.y:65.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
   65 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
/home/runner/work/reactos/reactos/src/dll/win32/msi/sql.y: warning: 1 reduce/reduce conflict [-Wconflicts-rr]
/home/runner/work/reactos/reactos/src/dll/win32/msi/sql.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
```
```
/home/runner/work/reactos/reactos/src/dll/win32/msxml3/xslpattern.y:73.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
   73 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
/home/runner/work/reactos/reactos/src/dll/win32/msxml3/xslpattern.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
```
```
/home/runner/work/reactos/reactos/src/dll/win32/vbscript/parser.y:84.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
   84 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
/home/runner/work/reactos/reactos/src/dll/win32/vbscript/parser.y: warning: 9 shift/reduce conflicts [-Wconflicts-sr]
/home/runner/work/reactos/reactos/src/dll/win32/vbscript/parser.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
[12098/12659] Generating wql.tab.c, wql.tab.h
/home/runner/work/reactos/reactos/src/dll/win32/wbemprox/wql.y:219.1-12: warning: deprecated directive: ‘%pure-parser’, use ‘%define api.pure’ [-Wdeprecated]
  219 | %pure-parser
      | ^~~~~~~~~~~~
      | %define api.pure
/home/runner/work/reactos/reactos/src/dll/win32/wbemprox/wql.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
```

**Comment by ColinFinck** [comment]
All build machines are running on RosBE 2.2.1 now.
I've used this PR as the testbed for RosBE 2.2.1 and it built fine everywhere.
Warnings can be fixed later, if at all (modified files need to be submitted to WINE and not sure they accept them there).

Let's fix the conflicts and finally get it done :)
@AmineKhaldi @Extravert-ir

**Comment by Extravert-ir** [comment]
Superseded by #3435

**Pull request closed without merging**